File: SantaClaus/SantaClaus-Cpp/santa.cpp
#include "santa.h"

#include "deer.h"
#include "elve.h"

#include <thread>
#include <iostream>

using namespace std;


void santa::deliver_presents()
{
    this_thread::sleep_for(chrono::milliseconds(m_dist(m_gen)));
}

void santa::solve_problems()
{
    this_thread::sleep_for(chrono::milliseconds(m_dist(m_gen)));
}

santa::santa(std::mutex *mt, std::condition_variable *cv)
    : m_gen(random_device{}())
    , m_dist(100, 5000)
    , mp_mt(mt)
    , mp_cv(cv)
{

}

void santa::operator()()
{
    cout << "Santa is asleep ..." << endl;
    while(true) {
        // This is considered as sleep time for Santa
        {
            unique_lock<mutex> guard(*mp_mt);

            // wait until other deers to arrive
            mp_cv->wait(guard, [&](){
                /*
                 * Unlike the statement in the little book of semaphores
                 * it is not possible for both of the conditions to happend
                 * at the same time
                 */
                return deer::all_deers_ready() || elve::problems_need_attention();
            });
        }



        // we could use if--else block since it is unlikely for both
        // to happen at the same time but for book comfortance we use
        // to ordered if block

        // Delivering presents is in the proiority
        if(deer::all_deers_ready()) {
            cout << "Santa is waked up to deliver presents: ";
            deliver_presents();
            deer::set_all_deers_free();
            cout << "done!" << endl;
            mp_cv->notify_all();
        }



        if(elve::problems_need_attention()) {
            cout << "Santa is waked up to solve elves problem: ";
            solve_problems();
            elve::set_all_problems_solved();
            cout << "done!" << endl;
            mp_cv->notify_all();
        }

        cout << "Santa is back to sleep ..." << endl;
    }
}

File: SantaClaus/SantaClaus-Go/santa.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type santa struct {
}

func makeSanta() santa {
	return santa{}
}

func (s santa) deliver_presents() {
	fmt.Println("Santa is delivering the presents")
	time.Sleep(time.Duration(rand.Intn(2000)) * time.Millisecond)
}

func (s santa) solve_problems() {
	fmt.Println("Santa is solving the problems")
	time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
}

func (s santa) run() {

	var deers []deer
	var elves []elve

	for {
		for len(deers) < 9 && len(elves) < 3 {
			select {
			case d := <-deers_returned:
				deers = append(deers, d)
			case e := <-elves_with_problems:
				elves = append(elves, e)
			}
		}

		if len(deers) == 9 {
			s.deliver_presents()

			for _, d := range deers {
				deers_on_vacation <- d
			}
			deers = deers[:0]
		}

		if len(elves) == 3 {
			s.solve_problems()

			for _, e := range elves {
				elves_making_toys <- e
			}
			elves = elves[:0]
		}
	}

}
